sound/soc/sof/sof-audio.c

```diff
@@ -11,6 +11,538 @@
+					  struct snd_soc_dapm_widget_list *list, int dir)
+{
+	struct snd_soc_dapm_widget *widget;
+	struct snd_soc_dapm_path *p;
+	int ret;
+	int i;
+
+	/*
+	 * Set up connections between widgets in the sink/source paths based on direction.
+	 * Some non-SOF widgets exist in topology either for compatibility or for the
+	 * purpose of connecting a pipeline from a host to a DAI in order to receive the DAPM
+	 * events. But they are not handled by the firmware. So ignore them.
+	 */
+	if (dir == SNDRV_PCM_STREAM_PLAYBACK) {
+		for_each_dapm_widgets(list, i, widget) {
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_sink_path(widget, p)
+				if (is_sof_widget(p->sink)) {
+					ret = sof_route_setup(sdev, widget, p->sink);
+					if (ret < 0)
+						return ret;
+				}
+		}
+	} else {
+		for_each_dapm_widgets(list, i, widget) {
+			if (!is_sof_widget(widget))
+				continue;
+
+			snd_soc_dapm_widget_for_each_source_path(widget, p)
+				if (is_sof_widget(p->source)) {
+					ret = sof_route_setup(sdev, p->source, widget);
+					if (ret < 0)
+						return ret;
+				}
+		}
+	}
+
+	return 0;
+}
+
+static struct snd_sof_widget *sof_find_pipeline_by_id(struct snd_sof_dev *sdev, int pipeline_id)
+{
+	struct snd_sof_widget *swidget;
+
+	list_for_each_entry(swidget, &sdev->widget_list, list)
+		if (swidget->id == snd_soc_dapm_scheduler && swidget->pipeline_id == pipeline_id)
+			return swidget;
+
+	return NULL;
+}
+
+int sof_widget_list_setup(struct snd_sof_dev *sdev, struct snd_sof_pcm *spcm, int dir)
+{
+	struct snd_soc_dapm_widget_list *list = spcm->stream[dir].list;
+	struct snd_soc_dapm_widget *widget;
+	struct snd_sof_widget *w, *_w;
+	int pipe_widget_count = 0;
+	int i, ret;
+	int j;
+
+	/* nothing to set up */
+	if (!list)
+		return 0;
+
+	/* set up widgets in the list */
+	for_each_dapm_widgets(list, i, widget) {
+		struct snd_sof_widget *swidget = widget->dobj.private;
+		struct snd_sof_widget *pipe_widget, *temp_pipe_widget;
+		bool pipeline_found = false;
+
+		if (!swidget)
+			continue;
+
+		ret = sof_widget_setup(sdev, swidget);
+		if (ret < 0)
+			goto widget_free;
+
+		/* find pipeline widget for the pipeline that this widget belongs to */
+		pipe_widget = sof_find_pipeline_by_id(sdev, swidget->pipeline_id);
+		if (!pipe_widget) {
+			ret = -EINVAL;
+			goto widget_free;
+		}
+
+		/* Add it to the spcm pipeline list if it isn't part of it already. */
+		list_for_each_entry(w, &spcm->stream[dir].pipeline_list, list) {
+			if (w->comp_id == pipe_widget->comp_id) {
+				pipeline_found = true;
+				break;
+			}
+		}
+		if (!pipeline_found) {
+			/*
+			 * Pipeline widgets can only be set up after the scheduling
+			 * widget has been set up. Create a temp widget to save it to the spcm
+			 * pipeline list as the original widget is already part of the core's
+			 * widget list.
+			 */
+			temp_pipe_widget = kmemdup(pipe_widget, sizeof(*pipe_widget), GFP_KERNEL);
```

Comment:
@ranj063 I also got this issue when I implemented ipc4.  kemedup is a heavy operation to me and it makes status duplicated and may mismatch. I suggest add a " struct list_head pipeline_list "  and link this to spcm pipeline_list. This also makes pipeline free simpler. It is common that one widget in multiple list.

Reply:
@RanderWang makes sense. lets go with a pointer to pipeline_list

Reply:
o, so list_head is not as normal pointer usage in C language.

Reply:
How many pcm will share one pipeline ?  Max = 8 is ok ?  can we define struct in spcm->stream like the following one to replace list_head ? 
````c
struct pipeline_array {
             int pipeline_count;
             struct snd_sof_widget *pipeline[8];
}
````